Accept index name in body on _close API

zembrzuski opened the pull request:
Relates to #82811

zembrzuski committed:
issue 82811 -Accept index name in body on _close API

elasticmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

nik9000 commented:
@lasticmachine, test this please

nik9000 commented:
Assigning to @DaveCTurner because it looks like he's talking about this already.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/indices/close/CloseIndexRequest.java:
```diff
@@ -114,4 +118,22 @@ public void writeTo(StreamOutput out) throws IOException {
+
+    public static CloseIndexRequest fromRestRequest(RestRequest request) throws IOException {
+        String[] indices = request.hasParam("index")
+            ? Strings.splitStringByCommaToArray(request.param("index"))
```
If a request mentions indices in both the URL and the body then we should definitely not silently ignore one of those collections. IMO we should reject requests that use both forms.

zembrzuski replied:
Ok. I added a better validation, and the requests are rejected when indices are specified both in path and body.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/indices/close/CloseIndexRequest.java:
```diff
@@ -114,4 +118,22 @@ public void writeTo(StreamOutput out) throws IOException {
+        return new CloseIndexRequest(indices);
+    }
+
+    private static String[] getIndicesFromBody(final RestRequest request) throws IOException {
```
Generally the `action` layer shouldn't know about REST things. Could we keep the parsing in the REST layer please?

zembrzuski replied:
Ok, the parsing is being executed in RestCloseIndexAction now.

DaveCTurner commented on server/src/test/java/org/elasticsearch/action/admin/indices/close/CloseIndexRequestTests.java:
```diff
@@ -11,14 +11,55 @@
 
 public class CloseIndexRequestTests extends ESTestCase {
 
+    public void testIndexPath() throws IOException {
```
I think we should have at least some YAML tests in `rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/indices.open/` to test this feature. I think we can probably move all of these tests there.

zembrzuski replied:
Ok, I added two more YAML tests.

DaveCTurner reviewed (commented):
Thanks @zembrzuski, I left a few initial comments.

zembrzuski committed:
code review #1

- rejecting requests when index is specified both in path and in body.
- added yaml tests
- request parsing moved to RestCloseIndexAction

zembrzuski commented:
Hi @DaveCTurner 
Just pushed another commit to try to address the recommended issues.
Thanks for reviewing this PR!

elasticmachine commented:
Pinging @elastic/clients-team (Team:Clients)

dakrone commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into issue_82811

dakrone commented:
@elasticmachine ok to test

sethmlarson commented:
Because this API is symmetrical with `open`, should we add support for both in the same PR?

dakrone commented:
@sethmlarson yes, I would prefer to support both so that they match.

@zembrzuski would you be able to add support for a body to the `/_open` API in this PR also?

sethmlarson commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/indices.open/10_basic.yml:
```diff
@@ -125,6 +125,67 @@
+      indices.close:
+        body:
+          index:
+            - "index_1"
```
Can we add a test case for sending a single string (ie `"index": "index_1") instead of a list as well? Thinking how clients implement path parameters for lists of values, we typically allow either a singular value or multiple values because "1" and ["1"] both serialize to the same path parameter value.

dakrone commented:
@zembrzuski are you still interested in this PR? Are you be able to address the feedback?

dakrone commented:
Going to close this one since it's a been a few months since it's had any update. Let me know if you're interested in revisiting it and we can reopen it.
